File: frontend/src/lib/project.ts
import api from './api';
import type { TaskCreate } from './task';

// Project interfaces
export interface Project {
  id: number;
  name: string;
  description?: string;
  owner_id: number;
  created_at: string;
  updated_at: string;
}

export interface ProjectCreate {
  name: string;
  description?: string;
  owner_id: number;
}

// Project API functions
export async function getProjects(): Promise<Project[]> {
  const response = await api.get('/projects/');
  return response.data;
}

export async function getProject(id: number): Promise<Project> {
  const response = await api.get(`/projects/${id}`);
  return response.data;
}

export async function createProject(data: ProjectCreate): Promise<Project> {
  const response = await api.post('/projects/', data);
  return response.data;
}

export async function updateProject(id: number, data: ProjectCreate): Promise<Project> {
  const response = await api.put(`/projects/${id}`, data);
  return response.data;
}

export async function deleteProject(id: number): Promise<void> {
  await api.delete(`/projects/${id}`);
}

// Additional project-specific endpoints
export async function getProjectTasks(projectId: number) {
  const response = await api.get(`/projects/${projectId}/tasks/`);
  return response.data;
}

export async function createProjectTask(projectId: number, data: TaskCreate) {
  const response = await api.post(`/projects/${projectId}/tasks/`, data);
  return response.data;
}

File: backend/app/api/project.py
from fastapi import APIRouter, Depends, HTTPException, Response
from sqlalchemy.orm import Session
from app.db.session import SessionLocal
from app.schemas.project import ProjectCreate, ProjectOut, ProjectUpdate
from app.schemas.task import TaskOut
from app.crud import project as project_crud
from app.core.deps import get_current_user, require_admin_or_scrum_master, get_db
from app.db.models import User
from typing import List

router = APIRouter(prefix="/projects", tags=["projects"])

@router.post("/", response_model=ProjectOut, status_code=201)
def create_project(
    project: ProjectCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_admin_or_scrum_master)
):
    return project_crud.create_project(db=db, project=project)

@router.get("/", response_model=List[ProjectOut])
def get_projects(
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    return project_crud.get_projects(db=db, skip=skip, limit=limit)

@router.get("/{project_id}", response_model=ProjectOut)
def get_project(
    project_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    project = project_crud.get_project_by_id(db=db, project_id=project_id)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    return project

@router.put("/{project_id}", response_model=ProjectOut)
def update_project(
    project_id: int,
    project: ProjectUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_admin_or_scrum_master)
):
    db_project = project_crud.get_project_by_id(db=db, project_id=project_id)
    if not db_project:
        raise HTTPException(status_code=404, detail="Project not found")
    return project_crud.update_project(db=db, db_project=db_project, updates=project)

@router.delete("/{project_id}", status_code=204)
def delete_project(
    project_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_admin_or_scrum_master)
):
    db_project = project_crud.get_project_by_id(db=db, project_id=project_id)
    if not db_project:
        raise HTTPException(status_code=404, detail="Project not found")
    project_crud.delete_project(db=db, db_project=db_project)
    return Response(status_code=204)

@router.get("/{project_id}/tasks", response_model=List[TaskOut])
def get_project_tasks(
    project_id: int,
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    return project_crud.get_project_tasks(db=db, project_id=project_id, skip=skip, limit=limit)

@router.get("/{project_id}/tasks/{task_id}", response_model=TaskOut)
def get_project_task(project_id: int, task_id: int, db: Session = Depends(get_db)):
    return project_crud.get_project_task(db=db, project_id=project_id, task_id=task_id)
